File: py/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if len(s) < 2:
            return len(s)

        map = {}
        left = 0
        ans = 0

        for i in range(len(s)):
            global left

            if s[i] in map and map[s[i]] >= left:
              left = map[s[i]] + 1

            map[s[i]] = i

            c = i - left + 1

            if c > ans:
                ans = c

        return ans

File: go/longest-substring-without-repeating-characters.go
package main

/*
解题方法：滑动窗口

时间复杂度：O(n)
*/
func lengthOfLongestSubstring(s string) int {
	left, maxLen := 0, 0
	// 使用slice比map更省内存；ascii范围 0 - 127
	// 提前申请长度，减少申请内存时间
	m := make([]int, 127)
	for right, c := range s {
		if m[c] != 0 {
			left = max(left, m[c])
		}
		maxLen = max(maxLen, right-left+1)
		m[c] = right + 1
	}
	return maxLen
}
